dll/win32/kernel32/kernel32_vista/vista.c

```diff
@@ -26,61 +26,61 @@
-                           LPWSTR lpExeName,
-                           PDWORD pdwSize)
+QueryFullProcessImageNameW(
+    _In_ HANDLE hProcess,
+    _In_ DWORD dwFlags,
+    _Out_writes_to_(*lpdwSize, *lpdwSize) LPWSTR lpExeName,
+    _Inout_ PDWORD lpdwSize)
 {
-    BYTE Buffer[sizeof(UNICODE_STRING) + MAX_PATH * sizeof(WCHAR)];
-    UNICODE_STRING *DynamicBuffer = NULL;
-    UNICODE_STRING *Result = NULL;
+    BOOL bRet;
+    DWORD dwBufferSize;
+    PUNICODE_STRING pBuffer;
     NTSTATUS Status;
-    DWORD Needed;
+    USHORT uCch;
 
-    Status = NtQueryInformationProcess(hProcess,
-                                       ProcessImageFileName,
-                                       Buffer,
-                                       sizeof(Buffer) - sizeof(WCHAR),
-                                       &Needed);
-    if (Status == STATUS_INFO_LENGTH_MISMATCH)
+    if (dwFlags && dwFlags != PROCESS_NAME_NATIVE)
     {
-        DynamicBuffer = RtlAllocateHeap(RtlGetProcessHeap(), 0, Needed + sizeof(WCHAR));
-        if (!DynamicBuffer)
-        {
-            BaseSetLastNTError(STATUS_NO_MEMORY);
-            return FALSE;
-        }
-
-        Status = NtQueryInformationProcess(hProcess,
-                                           ProcessImageFileName,
-                                           (LPBYTE)DynamicBuffer,
-                                           Needed,
-                                           &Needed);
-        Result = DynamicBuffer;
+        BaseSetLastNTError(STATUS_INVALID_PARAMETER_2);
+        return FALSE;
     }
-    else Result = (PUNICODE_STRING)Buffer;
-
-    if (!NT_SUCCESS(Status)) goto Cleanup;
 
-    if (Result->Length / sizeof(WCHAR) + 1 > *pdwSize)
+    dwBufferSize = *lpdwSize * sizeof(WCHAR) + sizeof(UNICODE_STRING);
+    pBuffer = (PUNICODE_STRING)RtlAllocateHeap(RtlGetProcessHeap(), 0, dwBufferSize);
+    if (!pBuffer)
     {
-        Status = STATUS_BUFFER_TOO_SMALL;
-        goto Cleanup;
+        BaseSetLastNTError(STATUS_NO_MEMORY);
+        return FALSE;
     }
 
-    *pdwSize = Result->Length / sizeof(WCHAR);
-    memcpy(lpExeName, Result->Buffer, Result->Length);
-    lpExeName[*pdwSize] = 0;
-
-Cleanup:
-    RtlFreeHeap(RtlGetProcessHeap(), 0, DynamicBuffer);
-
-    if (!NT_SUCCESS(Status))
+    bRet = FALSE;
+    Status = NtQueryInformationProcess(hProcess,
+                                       dwFlags ? ProcessImageFileName : ProcessImageFileNameWin32,
+                                       pBuffer,
+                                       dwBufferSize,
+                                       NULL);
+    if (Status == STATUS_INFO_LENGTH_MISMATCH)
+    {
+        BaseSetLastNTError(STATUS_BUFFER_TOO_SMALL);
+    } else if (!NT_SUCCESS(Status))
     {
         BaseSetLastNTError(Status);
+    } else
+    {
+        uCch = pBuffer->Length / sizeof(WCHAR);
+        if (uCch >= *lpdwSize)
+        {
+            BaseSetLastNTError(STATUS_BUFFER_TOO_SMALL);
```

Comment:
Set needed size too?

Reply:
No, `QueryFullProcessImageName` will not feedback required size.

Reply:
Right, tested by modules/rostests/winetests/kernel32/process.c .